### DISCUSSION TITLE
Guidelines on optimization/speed issues?

### DISCUSSION THREAD

**Discussion by Nightshade-tools** [libvips/libvips#2898] [Help wanted]
I haven't been able to find any guidelines or help on how to deal with speed issues in this library.
I have a long pipeline where I read in 16x images that together are over 150 000 x 150 000 pixels and I am not sure if I am doing things correctly or not to ensure that I run things at optimal speeds. Like should I save out the image after each processing step (and read pre-existing images from disk, to speed up future iterations and run-throughs of the pipeline?) - maybe I should never save out the image on disk?

My pipeline is as follows:
Read the 16 images > crop them (center 4 images gets skipped) > combine them into one image > resize down to 2000 pixels > save out. This process took a little under 1400 to 1600 seconds (23 to 26 minutes) and I don't know what step is taking so long! I have some similar code which only read and resize which runs on Pillow and Numpy +threading (Pool) and that code executes much, much faster: my PyVips code is 20x slower.

I'm using the PyVips bindings and I use the following Image functions:
new_from_file(), crop(), join(), resize(), pngsave()

Secondary questions:
Why does pngsave return a list[] - and what data does it hold? The pyvips docs are not clear on this)

General advice for working with GeoTifs? (16 bit TIFs)

**Comment by jcupitt** [comment]
Hi @Nightshade-tools,

I think you'll need to give some more information. What format are your input images, maybe a sample image, how are you opening them, etc. 

Could you copy-paste your code too? Otherwise to reproduce your problem I'll have to try to write something that does what I think you might be doing, and I'm bound to get it wrong.

**Reply by Nightshade-tools** [reply]
Hello,

Images are TIFF64, grayscale, 32/16 bit depth with LZW compression and with a resolution of 40 000 x 40 000.
The images I use are from the Copernicus height map data (EU-DEM): https://land.copernicus.eu/imagery-in-situ/eu-dem/eu-dem-v1.1
(Account is needed to download the tiles. Each tile is 40k by 40k and up to 4.9gb.

Using Image.signal_connect and an eval_handler function I have managed to pinpoint the problem (or part of the problem) to the image load:
No matter if I use Image.new_from_file() or Image.tiffload() load time is quite huge.

There's some older code in the project I am working on which use the Python tifffile, imread() and which performs edits on numpy arrays. The old imread+numpy code runs in about 1.5 minutes while doing the same thing with PyVips is about 7 times slower.

Here is a small example that you can run for comparison. I'm using 16x of those huge TIF files from Copernicus for this.


```
import numpy as np
from tifffile import imread
from pathlib import Path
import pyvips
from timeit import default_timer

def load_data(file_path_list):

    start = default_timer()
    for index, f in enumerate(file_path_list):
        # Case 1 - 70 seconds for one - 716 seconds for all
        image = pyvips.Image.new_from_file(f, access="sequential")

        # Case 2 - 68 seconds for one - 800 seconds for all
        # image = pyvips.Image.tiffload(f, access="sequential")

        # Case 3 - 25 seconds for one - 213 seconds for all
        # image_data = imread(f)  # Fastest - 12 to 18 seconds
        # np_image = np.array(image_data, dtype=np.float32)
        # image = pyvips.Image.new_from_array(np_image)

        image.avg()  # Just to run the pipeline

        # if index == 1:  # Early exit - if we just wanna try on N amount of Tifs
        #     break

    end = default_timer()
    print("Completed - Execution time: {0} seconds".format((end - start)))


if __name__ == '__main__':
    # Folder contains 16x TIF images of size 40 000 x 40 000 pixels each - downloaded from Copernicus
    some_path = Path("c:/Users/ user /Desktop/data/")
    tiff_image_list = list(some_path.rglob("*.TIF"))
    load_data(tiff_image_list)

```

Also, opening individual files with ImageMagick (in terminal/console) takes 30 seconds. I used this to inspect the source files (format, bitdepth, etc)

Is there any settings related function calls I can run at the start of my script to make PyVips faster (and/or use more resources). I have 128 gb of RAM at my disposalfor instance, and a 8x core intel i9 running at 3GHz.

**Comment by Nightshade-tools** [comment]
Another test case:
I found this on Stack Overflow:
https://stackoverflow.com/a/71588865

That person loads an even bigger tif with pyvips but the result is tremendously quicker - 3 seconds to open and then resize using thumbnail().

When I am running that very same logic inside the loop I posted above, processing those 16x big tif images, it takes me 2315 seconds - 38 minutes - (in average 144 seconds per image). This is 48x slower which suggests that maybe something else is wrong with my system and/or environment.

```
from pathlib import Path
import pyvips
from timeit import default_timer


def load_data(file_path_list):

    start = default_timer()
    for index, f in enumerate(file_path_list):
        out_file = pyvips.Image.thumbnail(f, 2048)
        out_name = f.name

        # Save with LZW compression
        out_file.tiffsave(some_path.joinpath("testfolder/{0}".format(out_name)), tile=True, compression='lzw')

    end = default_timer()
    print("Completed - Execution time: {0} seconds".format((end - start)))


if __name__ == '__main__':
    # Folder contains 16x TIF images of size 40 000 x 40 000 pixels each
    some_path = Path("c:/Users/ user /Desktop/")
    tiff_image_list = list(some_path.rglob("*.TIF"))
    load_data(tiff_image_list)
```

**Reply by jcupitt** [reply]
Yes, I regularly manipulate 300,000 x 300,000 pixel images without huge slowdowns.

I've downloaded a 4.9gb sample TIF from that site. They are 128x128 pixel tiled float LZW, which is a bit unusual.

As a quick benchmark I tried:

```
$ vips avg eu_dem_v11_E30N20.TIF
[geotiff warnings skipped]
-114491301794883564467991667206738935808.000000

real	1m1.059s
user	0m59.627s
sys	0m2.412s
```

So it's around 1m to decompress the whole 40k x 40k image, and there's no parallelism (the user time is the same as the real time).  libvips uses the libtiff LZW decoder, which runs inside the libtiff lock, so tile decode is single threaded. Maybe that's the issue?

I'll try joining 16 of these things up and timing it.

**Reply by Nightshade-tools** [reply]
I see, that's unfortunate.
I didn't know that libvips use libtiff for LZW decoding and that libtiff is single-threaded.
If I am not mistaken it seems like tifffile does multi-threaded tile/strip compression/decompression. Any plans on changing libtiff for tifffile?

**Reply by jcupitt** [reply]
You can do `.new_from_file("something.tif").copy_memory()` to read the whole image into RAM. You should be able to 16 of those in parallel, so single-threaded (per image) libtiff wouldn't matter.

**Comment by Nightshade-tools** [comment]
Hello,

I tried that with threading - see code below. Unfortunately I get an error:
multiprocessing.pool.MaybeEncodingError: Error sending result: '[<pyvips.Image 40000x40000 float, 1 bands, b-w>]'. Reason: 'TypeError("cannot pickle '_cffi_backend._CDataBase' object")'

I guess I need a different approach to threading

```
import logging.config
from multiprocessing import Pool, cpu_count
from pathlib import Path
import pyvips
from timeit import default_timer


def load_data(file_path_list):
    print("load_data()")

    start = default_timer()

    jobs = list(zip(file_path_list))
    threads = min(cpu_count(), 8)
    with Pool(threads) as pool:
        images = pool.starmap(load_data_parallel, jobs)

    end = default_timer()
    print("Completed - Execution time: {0} seconds".format((end - start)))
   
def load_data_parallel(file_path):
    image = pyvips.Image.new_from_file(file_path).copy_memory()
    image.avg()
    return image

```

**Reply by jcupitt** [reply]
Yes, multiprocessing will run each "thread" in a separate process and use pickle to serialize the results and pass them back. You'll need something more like true threads, or maybe shared memory segments.

**Reply by Nightshade-tools** [reply]
Thanks for the reply.

I believe I've found something. Is this along the lines of what you had in mind?

```
import logging.config
import threading
from pathlib import Path
import pyvips
from timeit import default_timer


def load_data(file_path_list):
    print("load_data()")
    start = default_timer()

    for f in file_path_list:
        image = pyvips.Image.new_from_file(f).copy_memory()
        image.avg()

    end = default_timer()
    print("Completed - Execution time: {0} seconds".format((end - start)))


def load_data_threaded(file_path_list):
    # print("load_data_threaded()")

    def load_data_parallel(file_path):
        # image = pyvips.Image.new_from_file(file_path).copy_memory()  # A tad bit slower actually
        image = pyvips.Image.new_from_file(file_path)
        image.avg()
        return image

    start = default_timer()

    threads = list()
    for index, f in enumerate(file_path_list):
        # print("Main: Create and start thread {0}".format(index))
        x = threading.Thread(target=load_data_parallel, args=(f,))
        threads.append(x)
        x.start()

    for index, thread in enumerate(threads):
        # print("Main: before joining thread {0}".format(index))
        thread.join()
        # print("Main: thread {0} done".format(index))

    end = default_timer()
    print("Completed - Execution time: {0} seconds".format((end - start)))


if __name__ == '__main__':
    pyvips_logger = logging.getLogger("pyvips")
    pyvips_logger.disabled = False  # Turns off warnings for the unknown tags

    # Folder contains 16x TIF images of size 40 000 x 40 000 pixels each
    some_path = Path("c:/Users/ user /Desktop/")
    tiff_image_list = list(some_path.rglob("*.TIF"))
    load_data_threaded(tiff_image_list)  # Approx 68 to 71 seconds without copy_memory() and ~73s with it
    # load_data(tiff_image_list)           # Approx 1770 seconds - 29 minutes
```


So looping through it all and creating the pyvips.Images +averaging them is incredibly slow - while the multi-threaded approach is over 25x faster.

**Reply by jcupitt** [reply]
Yes, that's the kind of thing. I think I've also found a cache sizing issue from looking at your files, give me a bit, I might have a patch.

**Comment by jcupitt** [comment]
Actually, it wasn't cache sizing, it was loaders failing to communicate seq mode up the pipeline if the loader supported random access. I've fixed it in git master libvips (a one line change!) and with this code:

```python
#!/usr/bin/python3

import sys
import pyvips

images = [pyvips.Image.new_from_file(filename, access="sequential")
          for filename in sys.argv[2:]]

joined = pyvips.Image.arrayjoin(images, across=4)

thumb = joined.thumbnail_image(2000)

thumb.write_to_file(sys.argv[1])
```

And this test dataset:

```
$ for i in {1..16}; do cp ../eu_dem_v11_E30N20.TIF $i.tif; done
```

(ie. the 40k x 40k elevation map copied 16 times) I now see:

```
$ VIPS_CONCURRENCY=1 VIPS_PROGRESS=1 time ./copernicus.py x.tif sample/*.tif
writing x.tif ...
copernicus.py temp-55: 2000 x 2000 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
copernicus.py temp-55: done in 376s          
354.99user 15.32system 6:16.16elapsed 98%CPU (0avgtext+0avgdata 6104700maxresident)k
42429760inputs+31280outputs (0major+1548308minor)pagefaults 0swaps
```

So that's about 6m of run time and 1.5gb of peak memory.

There's not a lot of parallelism (because libvips is still using the single-threaded libtiff decompressor), but this gets rid of the crazy overcomputation that was going on in the loader.

**Reply by jcupitt** [reply]
... I'll see if I can find a simple workaround for older libvipses.

**Reply by Nightshade-tools** [reply]
This is awesome John. Let me know when there's a new version including this in the Download area (releases) - I am not too familiar with creating my own builds/exes from source code unfortunately.

**Reply by jcupitt** [reply]
It'll be in 8.13, due out soon. There will be an -rc2 here this weekend, probably:

https://github.com/libvips/libvips/releases

**Reply by Nightshade-tools** [reply]
That's good news!

**Comment by jcupitt** [comment]
This loads the images well:

```python
#!/usr/bin/python3

import sys
import concurrent.futures as fut
import pyvips

def image_load(filename):
    image = pyvips.Image.new_from_file(filename, access="sequential")
    return image.copy_memory()

with fut.ThreadPoolExecutor() as executor:
    future_to_filename = {executor.submit(image_load, filename): filename
                          for filename in sys.argv[2:]}
    filename_to_image = {future_to_filename[future]: future.result()
                         for future in fut.as_completed(future_to_filename)}

images = [filename_to_image[filename] for filename in sys.argv[2:]]

joined = pyvips.Image.arrayjoin(images, across=4)

thumb = joined.thumbnail_image(2000)

thumb.write_to_file(sys.argv[1])
```

On this PC, it loads all 16 images in about 30s, but then seems to take over 2h to do the join! I'll see if I can see what's happening there.

**Reply by Nightshade-tools** [reply]
Hello again,
Yea I've noticed that doing operations on these images seem to be a bit slow but I've managed to make things a bit quicker using threading.

However now I notice that none of these processed images can be opened in an Image editor (Photoshop - or even Windows Photo Viewer). I couldn't open the original tiff files from Copernicus Land Service either, but thought it might be due to their sheer size.

I'm using pyvips.Image.tiffsave() after doing things like crop() and join() on the images and it appears that the saved tiff's are corrupt. Do you think this is an issue with pyvips (liptiff) or with the source files? (We have some older code that saves them out as PNG just fine).

**Reply by jcupitt** [reply]
Your images are tiled float TIFF, which very few packages can handle.

TIFF is a truly awful format, in that you can encode almost any image data as TIFF. Float images, plane separated, with 7:4:3 chroma subsampling? No problem! The problem is that making write very flexible makes read very difficult. I don't think anyone has ever written a program that can load any TIFF file.

If you convert them to something more common, like 16 bit unsigned, they should open OK. You might need to turn off tiling too.

**Reply by Nightshade-tools** [reply]
Thanks for your reply.
Would you say it would be wise for me to do the conversion to uint16 before doing any processing with pyvips?
I can do that in numpy on all the source data if it speeds things up.

**Comment by Nightshade-tools** [comment]
One thing before I am off for the weekend:
You mentioned a slowness with the join() method of Image. I've observed the same behavior with arrayjoin() - takes me approx 10 minutes to join those 16 big images.
I'm thinking of just making some function so I can do the join() stuff in parallel (threaded) to speed things up - but maybe that isn't necessary if this is related to the issue what you've found with join()?

**Reply by jcupitt** [reply]
That's right, arrayjoin followed by resize behaves puzzlingly poorly. I've not found a reason yet (trapped in stupid work issues).
